zebra-chain/src/addresses.rs

```diff
@@ -0,0 +1,222 @@
+    fn from(script: Script) -> Self {
+        TransparentAddress::PayToScriptHash {
+            network: Network::Mainnet,
+            script_hash: AddressPayloadHash::from(&script.0[..]),
+        }
+    }
+}
+
+impl From<PublicKey> for TransparentAddress {
+    fn from(pub_key: PublicKey) -> Self {
+        TransparentAddress::PayToPublicKeyHash {
+            network: Network::Mainnet,
+            pub_key_hash: AddressPayloadHash::from(&pub_key.serialize()[..]),
+        }
+    }
+}
+
+impl fmt::Debug for TransparentAddress {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut bytes = io::Cursor::new(Vec::new());
+        let _ = self.zcash_serialize(&mut bytes);
+
+        f.debug_tuple("TransparentAddress")
+            .field(&bs58::encode(bytes.get_ref()).with_check().into_string())
+            .finish()
+    }
+}
+
+impl ZcashSerialize for TransparentAddress {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        match self {
+            TransparentAddress::PayToScriptHash {
+                network,
+                script_hash,
+            } => {
+                if *network == Network::Mainnet {
+                    writer.write_all(&[0x1C, 0xBD][..])?
+                } else {
+                    // Dev network doesn't have a recommendation so we
+                    // default to testnet bytes.
+                    writer.write_all(&[0x1C, 0xBA][..])?
+                }
+                // XXX I'm asuming this is BigEndian because we're not
+                // explicitly making it LittleEndian?
+                writer.write_all(&script_hash.0)?
+            }
+            TransparentAddress::PayToPublicKeyHash {
+                network,
+                pub_key_hash,
+            } => {
+                if *network == Network::Mainnet {
+                    writer.write_all(&[0x1C, 0xB8][..])?
+                } else {
+                    // Dev network doesn't have a recommendation so we
+                    // default to testnet bytes.
+                    writer.write_all(&[0x1D, 0x25][..])?
+                }
+                // XXX I'm asuming this is BigEndian because we're not
+                // explicitly making it LittleEndian?
+                writer.write_all(&pub_key_hash.0)?
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl ZcashDeserialize for TransparentAddress {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        let mut version_bytes = [0; 2];
+        reader.read_exact(&mut version_bytes)?;
+
+        let mut hash_bytes = [0; 20];
+        reader.read_exact(&mut hash_bytes)?;
+
+        match version_bytes {
+            [0x1c, 0xbd] => Ok(TransparentAddress::PayToScriptHash {
+                network: Network::Mainnet,
+                script_hash: AddressPayloadHash(hash_bytes),
+            }),
+            // Currently only reading !mainnet versions as testnet.
```

Comment:
This comment is a little unclear to me -- from the spec it looks like that is the specific encoding for testnet addresses?

Reply:
This was because I know there is also dev network, but I don't know if they have different version bytes or whatever for their addresses, so maybe I should make a note about 'dev net collisions, beware'?

Reply:
Ah, yeah, I think that would be good to note, but maybe attach the note to the `_ =>` case so that it's clear that it's about possibly unhandled cases? The first time I read it I thought it meant that all non-mainnet versions were read as testnet, so I was confused because it didn't match (my reading of) the code.